File: PDFaCodigoJavaCSharpRustPython/MultipleChoiceExam/PythonCode/MultipleChoiceExam.py
import tkinter as tk
from tkinter import filedialog, messagebox
import random
import os

class MultipleChoiceExam(tk.Tk):
    def __init__(self):
        super().__init__()
        self.title("Multiple Choice Exam - No File")
        self.resizable(False, False)
        self.protocol("WM_DELETE_WINDOW", self.exitForm)

        self.headerFont = ("Arial", 18, "bold")
        self.examItemFont = ("Arial", 16, "bold")
        
        # Class level variables
        self.examTitle = ""
        self.header1 = ""
        self.header2 = ""
        self.numberTerms = 0
        self.term1 = [""] * 100
        self.term2 = [""] * 100
        self.numberTried = 0
        self.numberCorrect = 0
        self.correctAnswer = 0

        # Menu structure
        self.mainMenuBar = tk.Menu(self)
        self.config(menu=self.mainMenuBar)
        
        self.fileMenu = tk.Menu(self.mainMenuBar, tearoff=0)
        self.openMenuItem = self.fileMenu.add_command(label="Open", command=self.openMenuItemActionPerformed)
        self.fileMenu.add_separator()
        self.exitMenuItem = self.fileMenu.add_command(label="Exit", command=self.exitMenuItemActionPerformed)
        self.mainMenuBar.add_cascade(label="File", menu=self.fileMenu)

        self.optionsMenu = tk.Menu(self.mainMenuBar, tearoff=0)
        self.nameGroup = tk.IntVar(value=1)
        self.optionsMenu.add_radiobutton(label="Header 1", variable=self.nameGroup, value=1, command=self.header1MenuItemActionPerformed)
        self.optionsMenu.add_radiobutton(label="Header 2", variable=self.nameGroup, value=2, command=self.header2MenuItemActionPerformed)
        self.optionsMenu.add_separator()
        
        self.typeGroup = tk.IntVar(value=1)
        self.optionsMenu.add_radiobutton(label="Multiple Choice Answers", variable=self.typeGroup, value=1, command=self.mcMenuItemActionPerformed)
        self.optionsMenu.add_radiobutton(label="Type In Answers", variable=self.typeGroup, value=2, command=self.typeMenuItemActionPerformed)
        self.mainMenuBar.add_cascade(label="Options", menu=self.optionsMenu)

        # Controls
        self.headGivenLabel = tk.Label(self, font=self.headerFont, width=30)
        self.headGivenLabel.grid(row=0, column=0, pady=(10, 0), padx=10)

        self.givenLabel = tk.Label(self, font=self.examItemFont, bg="white", fg="blue", relief="solid", borderwidth=1, width=30, height=2)
        self.givenLabel.grid(row=1, column=0, padx=10)

        self.headAnswerLabel = tk.Label(self, font=self.headerFont, width=30)
        self.headAnswerLabel.grid(row=2, column=0, pady=(10, 0), padx=10)

        self.answerLabel = []
        for i in range(4):
            lbl = tk.Label(self, font=self.examItemFont, bg="white", fg="blue", relief="solid", borderwidth=1, width=30, height=2)
            lbl.grid(row=i+3, column=0, pady=(0, 10), padx=10)
            lbl.bind("<Button-1>", self.answerLabelMousePressed)
            self.answerLabel.append(lbl)

        self.answerTextField = tk.Entry(self, font=self.examItemFont, bg="white", fg="blue", width=30)
        self.answerTextField.bind("<Return>", self.answerTextFieldActionPerformed)
        # Se oculta inicialmente sacándolo del grid
        
        self.commentTextArea = tk.Text(self, font=("Courier New", 14, "bold italic"), bg="#FFFFC0", fg="red", width=40, height=3)
        self.commentTextArea.grid(row=7, column=0, pady=(0, 10), padx=10)
        self.commentTextArea.config(state=tk.DISABLED)

        self.nextButton = tk.Button(self, text="Next Question", command=self.nextButtonActionPerformed)
        self.nextButton.grid(row=8, column=0, pady=(0, 10))

        self.startButton = tk.Button(self, text="Start Exam", command=self.startButtonActionPerformed)
        self.startButton.grid(row=9, column=0, pady=(0, 10))

        self.centerWindow()

        # Initialize form
        self.startButton.config(state=tk.DISABLED)
        self.nextButton.config(state=tk.DISABLED)
        self.mainMenuBar.entryconfig("Options", state=tk.DISABLED)
        self.setCommentText(self.centerTextArea("Open Exam File to Start"))

    def centerWindow(self):
        self.update_idletasks()
        width = self.winfo_width()
        height = self.winfo_height()
        x = (self.winfo_screenwidth() // 2) - (width // 2)
        y = (self.winfo_screenheight() // 2) - (height // 2)
        self.geometry(f'{width}x{height}+{x}+{y}')

    def setCommentText(self, text):
        self.commentTextArea.config(state=tk.NORMAL)
        self.commentTextArea.delete(1.0, tk.END)
        self.commentTextArea.insert(tk.END, text)
        self.commentTextArea.config(state=tk.DISABLED)

    def exitForm(self):
        self.destroy()
    def exitMenuItemActionPerformed(self):
        self.destroy()

    def answerLabelMousePressed(self, event):
        if self.startButton.cget("text") == "Start Exam" or self.nextButton.cget("state") == tk.NORMAL:
            return
        
        labelSelected = self.answerLabel.index(event.widget)
        self.numberTried += 1
        correct = False

        if self.nameGroup.get() == 1:
            if self.answerLabel[labelSelected].cget("text") == self.term1[self.correctAnswer]:
                correct = True
        else:
            if self.answerLabel[labelSelected].cget("text") == self.term2[self.correctAnswer]:
                correct = True
                
        self.updateScore(correct)

    def answerTextFieldActionPerformed(self, event):
        if self.startButton.cget("text") == "Start Exam" or self.nextButton.cget("state") == tk.NORMAL:
            return
            
        self.answerTextField.config(state=tk.DISABLED)
        self.numberTried += 1
        ucTypedAnswer = self.answerTextField.get().upper()
        
        if self.nameGroup.get() == 1:
            ucAnswer = self.term1[self.correctAnswer].upper()
        else:
            ucAnswer = self.term2[self.correctAnswer].upper()
            
        correct = False
        if ucTypedAnswer == ucAnswer or self.soundex(ucTypedAnswer) == self.soundex(ucAnswer):
            correct = True
            
        self.updateScore(correct)

    def nextButtonActionPerformed(self):
        self.nextButton.config(state=tk.DISABLED)
        self.nextQuestion()

    def startButtonActionPerformed(self):
        if self.startButton.cget("text") == "Start Exam":
            self.startButton.config(text="Stop Exam")
            self.nextButton.config(state=tk.DISABLED)
            self.numberTried = 0
            self.numberCorrect = 0
            self.setCommentText("")
            self.mainMenuBar.entryconfig("File", state=tk.DISABLED)
            self.mainMenuBar.entryconfig("Options", state=tk.DISABLED)
            self.nextQuestion()
        else:
            self.startButton.config(text="Start Exam")
            self.nextButton.config(state=tk.DISABLED)
            if self.numberTried > 0:
                score = (self.numberCorrect / self.numberTried) * 100
                message = f"Questions Tried: {self.numberTried}\nQuestions Correct: {self.numberCorrect}\n\nYour Score: {score:.1f}%"
                messagebox.showinfo(f"{self.examTitle} Results", message)
            
            self.givenLabel.config(text="")
            for i in range(4): self.answerLabel[i].config(text="")
            self.answerTextField.config(state=tk.NORMAL)
            self.answerTextField.delete(0, tk.END)
            self.setCommentText(self.centerTextArea("Choose Options\nClick Start Exam"))
            self.mainMenuBar.entryconfig("File", state=tk.NORMAL)
            self.mainMenuBar.entryconfig("Options", state=tk.NORMAL)

    def openMenuItemActionPerformed(self):
        filepath = filedialog.askopenfilename(title="Open Exam File", filetypes=(("Exam Files", "*.csv"), ("All Files", "*.*")))
        if not filepath: return
        
        try:
            with open(filepath, 'r', encoding='utf-8') as inputFile:
                lines = [line.strip() for line in inputFile.readlines() if line.strip()]
            
            self.examTitle = self.parseLeft(lines[0])
            self.header1 = self.parseLeft(lines[1])
            self.header2 = self.parseRight(lines[1])
            self.numberTerms = 0
            
            for i in range(2, len(lines)):
                if self.numberTerms >= 100: break
                self.numberTerms += 1
                self.term1[self.numberTerms - 1] = self.parseLeft(lines[i])
                self.term2[self.numberTerms - 1] = self.parseRight(lines[i])
                
            if self.numberTerms < 5:
                messagebox.showerror("Exam File Error", "Must have at least 5 entries in exam file.")
                return
                
            self.title(f"Multiple Choice Exam - {self.examTitle}")
            
            if self.nameGroup.get() == 1:
                self.headGivenLabel.config(text=self.header2)
                self.headAnswerLabel.config(text=self.header1)
            else:
                self.headGivenLabel.config(text=self.header1)
                self.headAnswerLabel.config(text=self.header2)
                
            self.startButton.config(state=tk.NORMAL)
            self.mainMenuBar.entryconfig("Options", state=tk.NORMAL)
            self.setCommentText(self.centerTextArea("File Loaded, Choose Options\nClick Start Exam"))
            
        except Exception as e:
            messagebox.showerror("Multiple Choice Exam File Error", "Error reading in input file - make sure file is correct format.")

    def header1MenuItemActionPerformed(self):
        self.headGivenLabel.config(text=self.header2)
        self.headAnswerLabel.config(text=self.header1)

    def header2MenuItemActionPerformed(self):
        self.headGivenLabel.config(text=self.header1)
        self.headAnswerLabel.config(text=self.header2)

    def mcMenuItemActionPerformed(self):
        for i in range(4): self.answerLabel[i].grid(row=i+3, column=0, pady=(0, 10), padx=10)
        self.answerTextField.grid_forget()

    def typeMenuItemActionPerformed(self):
        for i in range(4): self.answerLabel[i].grid_forget()
        self.answerTextField.grid(row=3, column=0, pady=(0, 10), padx=10)

    def parseLeft(self, s):
        cl = s.find(",")
        return s[:cl] if cl != -1 else s

    def parseRight(self, s):
        cl = s.find(",")
        return s[cl+1:] if cl != -1 else ""

    def centerTextArea(self, s):
        charsPerLine = 33
        j = s.find("\n")
        if j == -1:
            return "\n" + self.spacePadding((charsPerLine - len(s)) // 2) + s
        else:
            l1 = s[:j]
            l2 = s[j+1:]
            return "\n" + self.spacePadding((charsPerLine - len(l1)) // 2) + l1 + "\n" + self.spacePadding((charsPerLine - len(l2)) // 2) + l2

    def spacePadding(self, n):
        return " " * max(0, n)

    def nextQuestion(self):
        termUsed = [False] * self.numberTerms
        index = [0] * 4
        self.setCommentText("")
        
        self.correctAnswer = random.randint(0, self.numberTerms - 1)
        
        if self.nameGroup.get() == 1:
            self.givenLabel.config(text=self.term2[self.correctAnswer])
        else:
            self.givenLabel.config(text=self.term1[self.correctAnswer])
            
        if self.typeGroup.get() == 1:
            for i in range(4):
                while True:
                    j = random.randint(0, self.numberTerms - 1)
                    if not termUsed[j] and j != self.correctAnswer:
                        break
                termUsed[j] = True
                index[i] = j
                
            index[random.randint(0, 3)] = self.correctAnswer
            
            for i in range(4):
                val = self.term1[index[i]] if self.nameGroup.get() == 1 else self.term2[index[i]]
                self.answerLabel[i].config(text=val)
        else:
            self.answerTextField.config(state=tk.NORMAL)
            self.answerTextField.delete(0, tk.END)
            self.answerTextField.focus_set()

    def updateScore(self, correct):
        if correct:
            self.numberCorrect += 1
            self.setCommentText(self.centerTextArea("Correct!"))
        else:
            self.setCommentText(self.centerTextArea("Sorry ... Correct Answer Shown"))
            
        if self.typeGroup.get() == 1:
            ans = self.term1[self.correctAnswer] if self.nameGroup.get() == 1 else self.term2[self.correctAnswer]
            self.answerLabel[0].config(text=ans)
            for i in range(1, 4): self.answerLabel[i].config(text="")
        else:
            ans = self.term1[self.correctAnswer] if self.nameGroup.get() == 1 else self.term2[self.correctAnswer]
            self.answerTextField.config(state=tk.NORMAL)
            self.answerTextField.delete(0, tk.END)
            self.answerTextField.insert(0, ans)
            
        self.startButton.config(state=tk.NORMAL)
        self.nextButton.config(state=tk.NORMAL)
        self.nextButton.focus_set()

    def soundex(self, w):
        wSound = [0, 1, 2, 3, 0, 1, 2, 0, 0, 2, 2, 4, 5, 5, 0, 1, 2, 6, 2, 3, 0, 1, 0, 2, 0, 2]
        wTemp = w.upper()
        l = len(w)
        s = ""
        
        if l != 0:
            s = wTemp[0]
            wPrev = 0
            if l > 1:
                for i in range(1, l):
                    cIndex = ord(wTemp[i]) - 65
                    if 0 <= cIndex <= 25:
                        wSnd = wSound[cIndex] + 48
                        if wSnd != 48 and wSnd != wPrev:
                            s += chr(wSnd)
                        wPrev = wSnd
        return s

if __name__ == "__main__":
    app = MultipleChoiceExam()
    app.mainloop()

File: PDFaCodigoJavaCSharpRustPython/MultipleChoiceExam/JavaCode/src/MultipleChoiceExam/MultipleChoiceExam.java
package	MultipleChoiceExam;

import javax.swing.filechooser.*;
import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.io.*;
import java.util.Random;
import java.text.*;

public class MultipleChoiceExam extends JFrame {
    
    JLabel headGivenLabel = new JLabel();
    JLabel givenLabel = new JLabel();
    JLabel headAnswerLabel = new JLabel();
    JLabel[] answerLabel = new JLabel[4];
    JTextField answerTextField = new JTextField();
    JTextArea commentTextArea = new JTextArea();
    JButton nextButton = new JButton();
    JButton startButton = new JButton();
    
    // menu structure
    JMenuBar mainMenuBar = new JMenuBar();
    JMenu fileMenu = new JMenu("File");
    JMenuItem openMenuItem = new JMenuItem("Open");
    JMenuItem exitMenuItem = new JMenuItem("Exit");
    JMenu optionsMenu = new JMenu("Options");
    JRadioButtonMenuItem header1MenuItem = new JRadioButtonMenuItem("Header 1", true);
    JRadioButtonMenuItem header2MenuItem = new JRadioButtonMenuItem("Header 2", false);
    JRadioButtonMenuItem mcMenuItem = new JRadioButtonMenuItem("Multiple Choice Answers", true);
    JRadioButtonMenuItem typeMenuItem = new JRadioButtonMenuItem("Type In Answers", false);
    ButtonGroup nameGroup = new ButtonGroup();
    ButtonGroup typeGroup = new ButtonGroup();
    
    Font headerFont = new Font("Arial", Font.BOLD, 18);
    Font examItemFont = new Font("Arial", Font.BOLD, 16);
    Dimension itemSize = new Dimension(370, 30);
    
    String examTitle;
    String header1, header2;
    int numberTerms;
    String[] term1 = new String[100];
    String[] term2 = new String[100];
    int numberTried, numberCorrect;
    int correctAnswer;
    Random myRandom = new Random();

    public static void main(String args[]) {
        // create frame
        new MultipleChoiceExam().show();
    }

    public MultipleChoiceExam() {
        // frame constructor
        setTitle("Multiple Choice Exam - No File");
        setResizable(false);
        addWindowListener(new WindowAdapter() {
            public void windowClosing(WindowEvent evt) {
                exitForm(evt);
            }
        });
        
        getContentPane().setLayout(new GridBagLayout());
        GridBagConstraints gridConstraints;
        
        headGivenLabel.setPreferredSize(itemSize);
        headGivenLabel.setFont(headerFont);
        gridConstraints = new GridBagConstraints();
        gridConstraints.gridx = 0;
        gridConstraints.gridy = 0;
        gridConstraints.insets = new Insets(10, 10, 0, 10);
        getContentPane().add(headGivenLabel, gridConstraints);
        
        givenLabel.setPreferredSize(itemSize);
        givenLabel.setFont(examItemFont);
        givenLabel.setBorder(BorderFactory.createLineBorder(Color.BLACK));
        givenLabel.setBackground(Color.WHITE);
        givenLabel.setForeground(Color.BLUE);
        givenLabel.setOpaque(true);
        givenLabel.setHorizontalAlignment(SwingConstants.CENTER);
        gridConstraints = new GridBagConstraints();
        gridConstraints.gridx = 0;
        gridConstraints.gridy = 1;
        gridConstraints.insets = new Insets(0, 10, 0, 10);
        getContentPane().add(givenLabel, gridConstraints);
        
        headAnswerLabel.setPreferredSize(itemSize);
        headAnswerLabel.setFont(headerFont);
        gridConstraints = new GridBagConstraints();
        gridConstraints.gridx = 0;
        gridConstraints.gridy = 2;
        gridConstraints.insets = new Insets(10, 10, 0, 10);
        getContentPane().add(headAnswerLabel, gridConstraints);
        
        for (int i = 0; i < 4; i++) {
            answerLabel[i] = new JLabel();
            answerLabel[i].setPreferredSize(itemSize);
            answerLabel[i].setFont(examItemFont);
            answerLabel[i].setBorder(BorderFactory.createLineBorder(Color.BLACK));
            answerLabel[i].setBackground(Color.WHITE);
            answerLabel[i].setForeground(Color.BLUE);
            answerLabel[i].setOpaque(true);
            answerLabel[i].setHorizontalAlignment(SwingConstants.CENTER);
            gridConstraints = new GridBagConstraints();
            gridConstraints.gridx = 0;
            gridConstraints.gridy = i + 3;
            gridConstraints.insets = new Insets(0, 10, 10, 10);
            getContentPane().add(answerLabel[i], gridConstraints);
            answerLabel[i].addMouseListener(new MouseAdapter() {
                public void mousePressed(MouseEvent e) {
                    answerLabelMousePressed(e);
                }
            });
        }
        
        answerTextField.setPreferredSize(itemSize);
        answerTextField.setFont(examItemFont);
        answerTextField.setBackground(Color.WHITE);
        answerTextField.setForeground(Color.BLUE);
        answerTextField.setVisible(false);
        gridConstraints = new GridBagConstraints();
        gridConstraints.gridx = 0;
        gridConstraints.gridy = 3;
        gridConstraints.insets = new Insets(0, 10, 10, 10);
        getContentPane().add(answerTextField, gridConstraints);
        answerTextField.addActionListener(new ActionListener () {
            public void actionPerformed(ActionEvent e) {
                answerTextFieldActionPerformed(e);
            }
        });
        
        commentTextArea.setPreferredSize(new Dimension(370, 80));
        commentTextArea.setFont(new Font("Courier New", Font.BOLD + Font.ITALIC, 18));
        commentTextArea.setBorder(BorderFactory.createLineBorder(Color.BLACK));
        commentTextArea.setEditable(false);
        commentTextArea.setBackground(new Color(255, 255, 192));
        commentTextArea.setForeground(Color.RED);
        gridConstraints = new GridBagConstraints();
        gridConstraints.gridx = 0;
        gridConstraints.gridy = 7;
        gridConstraints.insets = new Insets(0, 10, 10, 10);
        getContentPane().add(commentTextArea, gridConstraints);
        
        nextButton.setText("Next Question");
        gridConstraints = new GridBagConstraints();
        gridConstraints.gridx = 0;
        gridConstraints.gridy = 8;
        gridConstraints.insets = new Insets(0, 0, 10, 0);
        getContentPane().add(nextButton, gridConstraints);
        nextButton.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                nextButtonActionPerformed(e);
            }
        });
        
        startButton.setText("Start Exam");
        gridConstraints = new GridBagConstraints();
        gridConstraints.gridx = 0;
        gridConstraints.gridy = 9;
        gridConstraints.insets = new Insets(0, 0, 10, 0);
        getContentPane().add(startButton, gridConstraints);
        startButton.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                startButtonActionPerformed(e);
            }
        });
        
        // build menu structure
        setJMenuBar(mainMenuBar);
        mainMenuBar.add(fileMenu);
        fileMenu.add(openMenuItem);
        fileMenu.addSeparator();
        fileMenu.add(exitMenuItem);
        
        mainMenuBar.add(optionsMenu);
        optionsMenu.add(header1MenuItem);
        optionsMenu.add(header2MenuItem);
        optionsMenu.addSeparator();
        optionsMenu.add(mcMenuItem);
        optionsMenu.add(typeMenuItem);
        
        nameGroup.add(header1MenuItem);
        nameGroup.add(header2MenuItem);
        typeGroup.add(mcMenuItem);
        typeGroup.add(typeMenuItem);
        
        openMenuItem.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                openMenuItemActionPerformed(e);
            }
        });
        exitMenuItem.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                exitMenuItemActionPerformed(e);
            }
        });
        header1MenuItem.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                header1MenuItemActionPerformed(e);
            }
        });
        header2MenuItem.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                header2MenuItemActionPerformed(e);
            }
        });
        mcMenuItem.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                mcMenuItemActionPerformed(e);
            }
        });
        typeMenuItem.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                typeMenuItemActionPerformed(e);
            }
        });
        
        pack();
        Dimension screenSize = Toolkit.getDefaultToolkit().getScreenSize();
        // Nota: Se ha agregado el operador de multiplicación (*) que faltaba en el PDF para evitar errores de compilación
        setBounds((int) (0.5 * (screenSize.width - getWidth())), (int) (0.5 * (screenSize.height - getHeight())), getWidth(), getHeight());
        
        // initialize form
        startButton.setEnabled(false);
        nextButton.setEnabled(false);
        optionsMenu.setEnabled(false);
        commentTextArea.setText(centerTextArea("Open Exam File to Start"));
    }

    private void exitForm(WindowEvent evt) {
        System.exit(0);
    }

    private void answerLabelMousePressed(MouseEvent e) {
        boolean correct = false;
        int labelSelected;
        
        // make sure exam has started and question has not been answered
        if (startButton.getText().equals("Start Exam") || nextButton.isEnabled())
            return;
            
        // determine which label was clicked
        // get upper left corner of clicked label
        Point p = e.getComponent().getLocation();
        
        // determine index based on p
        for (labelSelected = 0; labelSelected < 20; labelSelected++) {
            if (p.x == answerLabel[labelSelected].getX() && p.y == answerLabel[labelSelected].getY()) 
                break;
        }
        
        // If already answered, exit
        numberTried++;
        if (header1MenuItem.isSelected()) {
            if (answerLabel[labelSelected].getText().equals(term1[correctAnswer])) 
                correct = true;
        } else {
            if (answerLabel[labelSelected].getText().equals(term2[correctAnswer])) 
                correct = true;
        }
        updateScore(correct);
    }

    private void answerTextFieldActionPerformed(ActionEvent e) {
        // Check type in answer
        boolean correct;
        String ucTypedAnswer, ucAnswer;
        
        // make sure exam has started and question has not been answered
        if (startButton.getText().equals("Start Exam") || nextButton.isEnabled()) 
            return;
            
        answerTextField.setEditable(false);
        numberTried++;
        ucTypedAnswer = answerTextField.getText().toUpperCase();
        
        if (header1MenuItem.isSelected())
            ucAnswer = term1[correctAnswer].toUpperCase();
        else
            ucAnswer = term2[correctAnswer].toUpperCase();
            
        correct = false;
        if (ucTypedAnswer.equals(ucAnswer) || soundex(ucTypedAnswer).equals(soundex(ucAnswer))) 
            correct = true;
            
        updateScore(correct);
    }

    private void nextButtonActionPerformed(ActionEvent e) {
        // Generate next question
        nextButton.setEnabled(false);
        nextQuestion();
    }

    private void startButtonActionPerformed(ActionEvent e) {
        String message;
        if (startButton.getText().equals("Start Exam")) {
            startButton.setText("Stop Exam");
            nextButton.setEnabled(false);
            // Reset the score
            numberTried = 0;
            numberCorrect = 0;
            commentTextArea.setText("");
            fileMenu.setEnabled(false);
            optionsMenu.setEnabled(false);
            nextQuestion();
        } else {
            startButton.setText("Start Exam");
            nextButton.setEnabled(false);
            if (numberTried > 0) {
                message = "Questions Tried: " + String.valueOf(numberTried) + "\n";
                message += "Questions Correct: " + String.valueOf(numberCorrect) + "\n\n";
                message += "Your Score: " + new DecimalFormat("0.0").format(100.0 * ((double) numberCorrect / numberTried)) + "%";
                JOptionPane.showConfirmDialog(null, message, examTitle + " Results", JOptionPane.DEFAULT_OPTION, JOptionPane.INFORMATION_MESSAGE);
            }
            givenLabel.setText("");
            answerLabel[0].setText("");
            answerLabel[1].setText("");
            answerLabel[2].setText("");
            answerLabel[3].setText("");
            answerTextField.setText("");
            commentTextArea.setText(centerTextArea("Choose Options\nClick Start Exam"));
            fileMenu.setEnabled(true);
            optionsMenu.setEnabled(true);
        }
    }

    private void openMenuItemActionPerformed(ActionEvent e) {
        String myLine;
        JFileChooser openChooser = new JFileChooser();
        openChooser.setDialogType(JFileChooser.OPEN_DIALOG);
        openChooser.setDialogTitle("Open Exam File");
        openChooser.addChoosableFileFilter(new FileNameExtensionFilter("Exam Files", "csv"));
        
        if (openChooser.showOpenDialog(this) == JFileChooser.APPROVE_OPTION) {
            try {
                BufferedReader inputFile = new BufferedReader(new FileReader(openChooser.getSelectedFile()));
                myLine = inputFile.readLine();
                examTitle = parseLeft(myLine);
                myLine = inputFile.readLine();
                header1 = parseLeft(myLine);
                header2 = parseRight(myLine);
                numberTerms = 0;
                
                do {
                    numberTerms++;
                    myLine = inputFile.readLine();
                    term1[numberTerms - 1] = parseLeft(myLine);
                    term2[numberTerms - 1] = parseRight(myLine);
                } while (inputFile.ready() && numberTerms < 100);
                
                if (numberTerms < 5) {
                    JOptionPane.showConfirmDialog(null, "Must have at least 5 entries in exam file.", "Exam File Error", JOptionPane.DEFAULT_OPTION, JOptionPane.ERROR_MESSAGE);
                    return;
                }
                
                inputFile.close();
                
                // establish frame title
                this.setTitle("Multiple Choice Exam - " + examTitle);
                
                // set up menu items
                header1MenuItem.setText(header1 + ", Given " + header2);
                header2MenuItem.setText(header2 + ", Given " + header1);
                
                if (header1MenuItem.isSelected()) {
                    headGivenLabel.setText(header2);
                    headAnswerLabel.setText(header1);
                } else {
                    headGivenLabel.setText(header1);
                    headAnswerLabel.setText(header2);
                }
                
                startButton.setEnabled(true);
                optionsMenu.setEnabled(true);
                commentTextArea.setText(centerTextArea("File Loaded, Choose Options\nClick Start Exam"));
                
            } catch (Exception ex) {
                JOptionPane.showConfirmDialog(null, "Error reading in input file - make sure file is correct format.", "Multiple Choice Exam File Error", JOptionPane.DEFAULT_OPTION, JOptionPane.ERROR_MESSAGE);
                return;
            }
        }
    }

    private void exitMenuItemActionPerformed(ActionEvent e) {
        System.exit(0);
    }

    private void header1MenuItemActionPerformed(ActionEvent e) {
        // Set up for naming header1, given header2
        headGivenLabel.setText(header2);
        headAnswerLabel.setText(header1);
    }

    private void header2MenuItemActionPerformed(ActionEvent e) {
        // Set up for naming header2, given header1
        headGivenLabel.setText(header1);
        headAnswerLabel.setText(header2);
    }

    private void mcMenuItemActionPerformed(ActionEvent e) {
        answerLabel[0].setVisible(true);
        answerLabel[1].setVisible(true);
        answerLabel[2].setVisible(true);
        answerLabel[3].setVisible(true);
        answerTextField.setVisible(false);
    }

    private void typeMenuItemActionPerformed(ActionEvent e) {
        answerLabel[0].setVisible(false);
        answerLabel[1].setVisible(false);
        answerLabel[2].setVisible(false);
        answerLabel[3].setVisible(false);
        answerTextField.setVisible(true);
    }

    private String parseLeft(String s) {
        int cl;
        // find comma
        cl = s.indexOf(",");
        return (s.substring(0, cl));
    }

    private String parseRight(String s) {
        int cl;
        // find comma
        cl = s.indexOf(",");
        return (s.substring(cl + 1));
    }

    private String centerTextArea(String s) {
        // centers up to two lines in text area
        int charsPerLine = 33;
        String sOut = "";
        int j = s.indexOf("\n");
        
        if (j == -1) {
            // single line
            sOut = "\n" + spacePadding((int) ((charsPerLine - s.length()) / 2)) + s;
        } else {
            // first line
            String l = s.substring(0, j);
            sOut = "\n" + spacePadding((int) ((charsPerLine - l.length()) / 2)) + l;
            // second line
            l = s.substring(j + 1);
            sOut += "\n" + spacePadding((int) ((charsPerLine - l.length()) / 2)) + l ;
        }
        return(sOut);
    }

    private String spacePadding(int n) {
        String s = "";
        if (n != 0)
            for (int i = 0; i < n; i++)
                s += " ";
        return(s);
    }

    private void nextQuestion() {
        boolean[] termUsed = new boolean[numberTerms];
        int[] index = new int[4];
        int j;
        commentTextArea.setText("");
        
        // Generate the next question based on selected options 
        correctAnswer = myRandom.nextInt(numberTerms);
        
        if (header1MenuItem.isSelected()) {
            givenLabel.setText(term2[correctAnswer]);
        } else {
            givenLabel.setText(term1[correctAnswer]);
        }
        
        if (mcMenuItem.isSelected()) {
            // Multiple choice answers
            for (int i = 0; i < numberTerms; i++) {
                termUsed[i] = false;
            }
            // Pick four random possiblities
            for (int i = 0; i < 4; i++) {
                do {
                    j = myRandom.nextInt(numberTerms);
                } while (termUsed[j] || j == correctAnswer);
                termUsed[j] = true;
                index[i] = j;
            }
            // Replace one with correct answer
            index[myRandom.nextInt(4)] = correctAnswer;
            
            // Display multiple choice answers in label boxes
            if (header1MenuItem.isSelected()) {
                answerLabel[0].setText(term1[index[0]]);
                answerLabel[1].setText(term1[index[1]]);
                answerLabel[2].setText(term1[index[2]]);
                answerLabel[3].setText(term1[index[3]]);
            } else {
                answerLabel[0].setText(term2[index[0]]);
                answerLabel[1].setText(term2[index[1]]);
                answerLabel[2].setText(term2[index[2]]);
                answerLabel[3].setText(term2[index[3]]);
            }
        } else {
            // Type-in answers
            answerTextField.setEditable(true);
            answerTextField.setText("");
            answerTextField.requestFocus();
        }
    }

    private void updateScore(boolean correct) {
        // Check if answer is correct
        if (correct) {
            numberCorrect++;
            commentTextArea.setText(centerTextArea("Correct!"));
        } else {
            commentTextArea.setText(centerTextArea("Sorry ... Correct Answer Shown"));
        }
        
        // Display correct answer
        if (mcMenuItem.isSelected()) {
            if (header1MenuItem.isSelected())
                answerLabel[0].setText(term1[correctAnswer]);
            else
                answerLabel[0].setText(term2[correctAnswer]);
                
            answerLabel[1].setText("");
            answerLabel[2].setText("");
            answerLabel[3].setText("");
        } else {
            if (header1MenuItem.isSelected())
                answerTextField.setText(term1[correctAnswer]);
            else
                answerTextField.setText(term2[correctAnswer]);
        }
        
        startButton.setEnabled(true);
        nextButton.setEnabled(true);
        nextButton.requestFocus();
    }

    public String soundex(String w) {
        // Generates Soundex code for W based on Unicode value
        // Allows answers whose spelling is close, but not exact 
        String wTemp, s = "";
        int l;
        int wPrev, wSnd, cIndex;
        
        // Load soundex function array 
        int[] wSound = {0, 1, 2, 3, 0, 1, 2, 0, 0, 2, 2, 4, 5, 5, 0, 1, 2, 6, 2, 3, 0, 1, 0, 2, 0, 2};
        wTemp = w.toUpperCase();
        l = w.length();
        
        if (l != 0) {
            s = String.valueOf(w.charAt(0));
            wPrev = 0;
            if (l > 1) {
                for (int i = 1; i < l; i++) {
                    cIndex = (int) wTemp.charAt(i) - 65;
                    if (cIndex >= 0 && cIndex <= 25) {
                        wSnd = wSound[cIndex] + 48;
                        if (wSnd != 48 && wSnd != wPrev) {
                            s += String.valueOf((char) wSnd);
                        }
                        wPrev = wSnd;
                    }
                }
            }
        } else {
            s = "";
        }
        return(s);
    }
}
